In this article I will examine how to look up nearby venues by using Foursquare API to retrieve data from Foursquare. 

I will extract different neighborhood in Munich, Germany for amenities. We will define the acceptable driving distance to be within 1.5 km (walking distance)

Based on definition of our problem, we must consider the following factors:

type and number of venues in the surrounding area of selected neighborhood 
It is better to use regularly spaced grid of locations, centered around city center, to set as a starting point to scrape the amenities in the selected neighborhood.

Following data sources will be needed to extract/generate the required information:

1. center of selected neiborhoods will be generated algorithmically and approximate addresses of centers of those areas will be obtained using Openstreet Maps API reverse geocoding
2. type and number of amenities and their type and location in every neighborhood will be obtained using Foursquare API


In [1]:
#import libraries
import pandas as pd
import numpy as np

import re
from geopy.geocoders import Nominatim
import shapely
import pyproj
import math
import folium
import requests
import pandas as pd
import time
import numpy as np
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors

import warnings
warnings.filterwarnings('ignore')

In [6]:
##### Define the loation of interest:
within_km = 1.5 # Select the within range in km (walking distance?)
bubble_detail_factor = 7 #varible factor determining how big the bubble will be, higher value = smaller bubble

# Read in the addresses of intrested locations in to a dataframe
d = {'listed_location': ['Altstadt-Lehel, Munich', 
                         'Ludwigsvorstadt and Isarvorstadt, Munich',
                         'Maxvorstadt, Munich',
                         'Schwabing West, Munich',
                         'Au-Haidhausen, Munich',
                         'Sendling, Munich',
                         'Sendling – Westpark, Munich',
                         'Schwanthalerhöhe, Munich', 
                         'Neuhausen Nymphenburg, Munich',
                         'Moosach, Munich',
                         'Milbertshofen und Am Hart, Munich',
                         'Schwabing-Freimann, Munich',
                         'Bogenhausen, Munich',
                         'Berg am Laim, Munich',
                         'Trudering – Riem, Munich',
                         'Ramersdorf und Perlach, Munich',
                         'Obergiesing, Munich',
                         'Untergiesing und Harlaching, Munich',
                         'Thalkirchen, Munich',
                         'Obersendling, Munich',
                         'Forstenried, Munich',
                         'Fürstenried-Solln, Munich',
                         'Hadern, Munich',
                         'Pasing – Obermenzing, Munich',
                         'Aubing-Lochhausen-Langwied, Munich',
                         'Allach Untermenzing, Munich',
                         'Feldmoching-Hasenbergl, Munich',
                         'Laim, Munich',                       
                        ]}
dfListing = pd.DataFrame(data=d)
dfListing

listed_location
0                     Altstadt-Lehel, Munich
1   Ludwigsvorstadt and Isarvorstadt, Munich
2                        Maxvorstadt, Munich
3                     Schwabing West, Munich
4                      Au-Haidhausen, Munich
5                           Sendling, Munich
6                Sendling – Westpark, Munich
7                   Schwanthalerhöhe, Munich
8              Neuhausen Nymphenburg, Munich
9                            Moosach, Munich
10         Milbertshofen und Am Hart, Munich
11                Schwabing-Freimann, Munich
12                       Bogenhausen, Munich
13                      Berg am Laim, Munich
14                  Trudering – Riem, Munich
15            Ramersdorf und Perlach, Munich
16                       Obergiesing, Munich
17       Untergiesing und Harlaching, Munich
18                       Thalkirchen, Munich
19                      Obersendling, Munich
20                       Forstenried, Munich
21                 Fürstenried-Solln, Munich
22                            Hadern, Munich
23              Pasing – Obermenzing, Munich
24        Aubing-Lochhausen-Langwied, Munich
25               Allach Untermenzing, Munich
26            Feldmoching-Hasenbergl, Munich
27                              Laim, Munich

Munich is located in Upper Bavaria, one of the 7 administrative districts of Bavaria. and is the capital of the administrative district of Upper Bavaria. Munich is also the capital of the Federal State and Free State of Bavaria. The city is the third largest city in Germany after Berlin and Hamburg. Since 2015, the population of Munich has exceeded 1.5 million for the first time. At the beginning of 2017, the Bavarian metropolis has 1.54 million inhabitants. So the population is growing rapidly, which leads to a housing shortage. The city has the highest rents and property prices in Germany.
The city of Munich is politically and administratively divided into 27 districts. The names of the districts are largely familiar to Munich’s residents, as they usually correspond to the traditional names of the city districts. The first district is the Altstadt (Old Town), the districts around the city centre have low numbers. The high double-digit districts are the suburbs, which are located on the outskirts of the city. The numbers of the districts are rather unknown in Munich.

Reading in the API keys and client ID for both Google and Foursquare in to variables:

In [63]:
#Read in API keys from a seperate text file, and define the version of the Foursquare API
behdad = open("behdad.txt", "r")
foursquare_client_id = re.search('(?<=Foursquare CLIENT_ID:)\S+',re.findall(r'Foursquare CLIENT_ID:.*', behdad.read())[0])[0]
behdad.seek(0)
foursquare_client_secret = re.search('(?<=Foursquare CLIENT_SECRET:)\S+',re.findall(r'Foursquare CLIENT_SECRET:.*', behdad.read())[0])[0]
behdad.seek(0)

VERSION = '20180323' # Foursquare API version
LIMIT = 7 # A default Foursquare API limit value

## Neighborhood Candidates
For this project, we would need to select centroids of latitude & longitude coordinates to scan nearby areas. The method will be creating grids of cells within 1.5 km of our selected addresses, which is approximate 3 x 3 km centered around the selected address.

Lets first find the latitude & longitude of the select city, using Openstreet Maps geocoding API.

In [8]:
geolocator = Nominatim(user_agent="foursqure_agent")

def get_coordinates(listing):
    try:
        geocode_result = geolocator.geocode(listing)
    except IndexError:
        print("Address was wrong...")
    except Exception as e:
        print("Unexpected error occurred.", e )
    return geocode_result[1]

Now let's create a grid of area candidates, equally spaced, centered around the given property address and within ~0.2143km (within 1.5 km divided by bubble size factor of 7) from the center. Each grid "bubble" surrounding the address will be defined as circular areas with a radius of 0.75 km (within 1.5 km, there for the radius is half of that), so grid bubbles' center will be 0.4286 km apart.

Using a method already available online, to accurately calculate distances we need to create our grid of locations in Cartesian 2D coordinate system which allows us to calculate distances in meters (not in latitude/longitude degrees). Then we will project those coordinates back to latitude/longitude degrees to be shown on Folium map. So let's create functions to convert between WGS84 spherical coordinate system (latitude/longitude degrees) and UTM Cartesian coordinate system (X/Y coordinates in meters).

UTM stands for "Universal Transverse Mercator".
Reference: [Cartesian/Projected Coordinate Systems, UTM](https://openpress.usask.ca/introgeomatics/chapter/cartesianprojected-coordinate-systems-utm/#:~:text=Universal%20Transverse%20Mercator%20(UTM)%20is,in%20metric%20units%20(metres).&text=Conversions%20from%20one%20coordinate%20system,a%20mathematical%20process%20called%20projection.)

In [7]:
#longitude & latitude to UTM xy
def lonlat_to_xy(lon, lat):
    proj_latlon = pyproj.Proj(proj='latlong',datum='WGS84')
    proj_xy = pyproj.Proj(proj="utm", zone=33, datum='WGS84')
    xy = pyproj.transformer.transform(proj_latlon, proj_xy, lon, lat)
    return xy[0], xy[1]

#UTM xy to longitude & latitude
def xy_to_lonlat(x, y):
    proj_latlon = pyproj.Proj(proj='latlong',datum='WGS84')
    proj_xy = pyproj.Proj(proj="utm", zone=33, datum='WGS84')
    lonlat = pyproj.transformer.transform(proj_xy, proj_latlon, x, y)
    return lonlat[0], lonlat[1]

#calculate distance between the two points (x1, y1) and (x2, y2)
def calc_xy_distance(x1, y1, x2, y2):
    dx = x2 - x1
    dy = y2 - y1
    return math.sqrt(dx*dx + dy*dy)

The very next step is to create a **hexagonal grid of cells**: we offset every other row, and adjust vertical row spacing so that **every cell center is equally distant from all it's neighbors**.

For more details, please refer to: [Hexagonal Grids](https://www.redblobgames.com/grids/hexagons/#:~:text=Size%20and%20Spacing%23&text=In%20the%20pointy%20orientation%2C%20a,from%20sin(60%C2%B0).&text=The%20horizontal%20distance%20between%20adjacent,is%20h%20*%203%2F4%20.)

Applying the get_coordinates() function and adding to the results to the dataframe

In [10]:
dfListing['location_corr'] = dfListing['listed_location'].apply(get_coordinates)
dfListing

listed_location  \
0                     Altstadt-Lehel, Munich   
1   Ludwigsvorstadt and Isarvorstadt, Munich   
2                        Maxvorstadt, Munich   
3                     Schwabing West, Munich   
4                      Au-Haidhausen, Munich   
5                           Sendling, Munich   
6                Sendling – Westpark, Munich   
7                   Schwanthalerhöhe, Munich   
8              Neuhausen Nymphenburg, Munich   
9                            Moosach, Munich   
10         Milbertshofen und Am Hart, Munich   
11                Schwabing-Freimann, Munich   
12                       Bogenhausen, Munich   
13                      Berg am Laim, Munich   
14                  Trudering – Riem, Munich   
15            Ramersdorf und Perlach, Munich   
16                       Obergiesing, Munich   
17       Untergiesing und Harlaching, Munich   
18                       Thalkirchen, Munich   
19                      Obersendling, Munich   
20                       Forstenried, Munich   
21                 Fürstenried-Solln, Munich   
22                            Hadern, Munich   
23              Pasing – Obermenzing, Munich   
24        Aubing-Lochhausen-Langwied, Munich   
25               Allach Untermenzing, Munich   
26            Feldmoching-Hasenbergl, Munich   
27                              Laim, Munich   

                               location_corr  
0                   (48.1378285, 11.5745823)  
1                   (48.1303398, 11.5733658)  
2                   (48.1510916, 11.5624179)  
3                   (48.1682709, 11.5698727)  
4    (48.130273849999995, 11.59833361534854)  
5                   (48.1180125, 11.5390832)  
6          (48.11803085, 11.519332770284128)  
7                   (48.1337822, 11.5410566)  
8                   (48.1542217, 11.5315172)  
9                   (48.1798949, 11.5105712)  
10                  (48.1823848, 11.5750432)  
11           (48.1892784, 11.60858258301819)  
12                  (48.1547823, 11.6334838)  
13                  (48.1234833, 11.6334511)  
14                  (48.1231749, 11.6640782)  
15  (48.100894249999996, 11.633370907753356)  
16         (48.10393445, 11.590683288923096)  
17                  (48.1149632, 11.5701894)  
18                  (48.1028401, 11.5459789)  
19                  (48.0983636, 11.5380152)  
20                  (48.0847807, 11.4947764)  
21                  (48.0889107, 11.4804702)  
22                   (48.118064, 11.4818417)  
23         (48.15236275, 11.468434085049713)  
24          (48.16505875, 11.40022147296721)  
25         (48.19515735, 11.462973336774297)  
26         (48.21846185, 11.520409083159603)  
27                  (48.1395506, 11.5021658)

In [11]:
#mapping out the locations and their surrounding venues
neighborhood=[]
latitudes = []
longitudes = []
distances_from_center = []
xs = []
ys = []
map_city = folium.Map(location=dfListing['location_corr'][1], zoom_start=(11))

for (index, row) in (dfListing.iterrows()):
    listed_location_x, listed_location_y = lonlat_to_xy(row ['location_corr'][1], row ['location_corr'][0]) 
    folium.Marker(row['location_corr'], popup=row['listed_location']).add_to(map_city)
    m = within_km * 1000
    k = math.sqrt(3) / 2 # Vertical offset for hexagonal grid cells
    x_min = listed_location_x - m
    x_step = m/(bubble_detail_factor/2)
    y_min = listed_location_y - m - (int(21/k)*k*m/bubble_detail_factor - 2*m)/2
    y_step = m/(bubble_detail_factor/2) * k
    for i in range(0, int(21/k)):
        y = y_min + i * y_step
        x_offset = (m/bubble_detail_factor) if i%2==0 else 0
        for j in range(0, 21):
            x = x_min + j * x_step + x_offset
            distance_from_center = calc_xy_distance(listed_location_x, listed_location_y, x, y)
            if (distance_from_center <= (m+1)):
                lon, lat = xy_to_lonlat(x, y)
                neighborhood.append(row['listed_location'])
                latitudes.append(lat)
                longitudes.append(lon)
                distances_from_center.append(distance_from_center)
                xs.append(x)
                ys.append(y)    

In [12]:
for lat, lon in zip(latitudes, longitudes):
    #folium.CircleMarker([lat, lon], radius=2, color='blue', fill=True, fill_color='blue', fill_opacity=1).add_to(map_city) 
    folium.Circle([lat, lon], radius=m/bubble_detail_factor, color='blue', fill=False).add_to(map_city)
    #folium.Marker([lat, lon]).add_to(map_city)
map_city

In [22]:
geolocator = Nominatim(user_agent="foursqure_agent")

def get_address(latlong):
    try:
        geocode_result = geolocator.reverse(latlong)
    except IndexError:
        print("latlong was wrong...")
    except Exception as e:
        print("Unexpected error occurred.", e )
    return geocode_result[0]

In [23]:
print('Obtaining location addresses: ', end='')
addresses = []
for latlon in zip(latitudes, longitudes):
    address = get_address(latlon)
    if address is None:
        address = 'NO ADDRESS'
    addresses.append(address)
    print(' .', end='')

Obtaining location addresses:  . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . .

In [24]:
df_locations = pd.DataFrame({'Neighborhood' : neighborhood,
                             'Address': addresses, #address of each grid bubble
                             'Latitude': latitudes,
                             'Longitude': longitudes,
                             'X': xs,
                             'Y': ys,
                             'Distance from center': distances_from_center})

df_locations

Neighborhood  \
0     Altstadt-Lehel, Munich   
1     Altstadt-Lehel, Munich   
2     Altstadt-Lehel, Munich   
3     Altstadt-Lehel, Munich   
4     Altstadt-Lehel, Munich   
...                      ...   
1199            Laim, Munich   
1200            Laim, Munich   
1201            Laim, Munich   
1202            Laim, Munich   
1203            Laim, Munich   

                                                Address   Latitude  Longitude  \
0     24, Eduard-Schmid-Straße, Untere Au, Bezirkste...  48.124495  11.575470   
1     58, Pestalozzistraße, Bezirksteil Glockenbach,...  48.127571  11.566624   
2     35, Westermühlstraße, Bezirksteil Glockenbach,...  48.127743  11.572374   
3     Erhardtstraße, Bezirksteil Deutsches Museum, L...  48.127914  11.578123   
4     45, Zeppelinstraße, Untere Au, Bezirksteil Unt...  48.128086  11.583872   
...                                                 ...        ...        ...   
1199  An der Schloßmauer, Bezirksteil Nymphenburg, N...  48.149286  11.492859   
1200  Zuccalistraße, Bezirksteil Nymphenburg, Neuhau...  48.149462  11.498610   
1201  30, Wotanstraße, Bezirksteil Nymphenburg, Neuh...  48.149637  11.504361   
1202  Königlicher Hirschgarten, Hirschgarten, Birket...  48.149812  11.510112   
1203  57, Richildenstraße, Bezirksteil Nymphenburg, ...  48.152883  11.501259   

                  X             Y  Distance from center  
0     245177.072223  5.335811e+06           1484.614978  
1     244534.215080  5.336182e+06           1285.714286  
2     244962.786509  5.336182e+06           1133.893419  
3     245391.357937  5.336182e+06           1133.893419  
4     245819.929366  5.336182e+06           1285.714286  
...             ...           ...                   ...  
1199  239156.457330  5.338843e+06           1285.714286  
1200  239585.028758  5.338843e+06           1133.893419  
1201  240013.600187  5.338843e+06           1133.893419  
1202  240442.171616  5.338843e+06           1285.714286  
1203  239799.314473  5.339214e+06           1484.614978  

[1204 rows x 7 columns]

So, we have 1204 grid bubbles for us to look at. Next is to use Foursquare to scrape all the venues within each grid bubbles (if any):

Next, we are going to start utilizing the Foursquare API to explore the addresses and segment them.

In [64]:
def getNearbyVenues(names, latitudes, longitudes, neighborhood, radius=m/20):
    
    venues_list=[]
    counter = 0
    for name, lat, lng, neighborhood in zip(names, latitudes, longitudes, neighborhood):
        print(neighborhood, '->', name) #to make sure the programming is running and at which stage
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
                foursquare_client_id, 
                foursquare_client_secret, 
                VERSION, 
                lat, 
                lng, 
                radius, 
                LIMIT)
        # make the GET request
        try:
            results = requests.get(url).json()["response"]['groups'][0]['items']
        except requests.exceptions.RequestException as err:
            print ("OOps: Something Else",err)
        except requests.exceptions.HTTPError as errh:
            print ("Http Error:",errh)
        except requests.exceptions.ConnectionError as errc:
            print ("Error Connecting:",errc)
        except requests.exceptions.Timeout as errt:
            print ("Timeout Error:",errt)  

        # return only relevant information for each nearby venue
        venues_list.append([(
                neighborhood,
                name, 
                lat, 
                lng, 
                v['venue']['id'],
                v['venue']['name'], 
                v['venue']['location']['lat'], 
                v['venue']['location']['lng'],  
                v['venue']['location']['formattedAddress'][0],
                v['venue']['categories'][0]['name']) for v in results])
            
    print(venues_list)
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    print(nearby_venues)
    nearby_venues.columns = ['Neighborhood',
                             'District', 
                  'District Latitude', 
                  'District Longitude', 
                  'id',
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Address',
                  'Venue Category']
    
    return(nearby_venues)

city_venues = getNearbyVenues(names=df_locations['Address'],
                                   latitudes=df_locations['Latitude'],
                                   longitudes=df_locations['Longitude'],
                                  neighborhood = df_locations['Neighborhood']
                                  )

Altstadt-Lehel, Munich -> 24, Eduard-Schmid-Straße, Untere Au, Bezirksteil Untere Au, Au, Au-Haidhausen, München, Bayern, 81541, Deutschland
Altstadt-Lehel, Munich -> 58, Pestalozzistraße, Bezirksteil Glockenbach, Ludwigsvorstadt-Isarvorstadt, München, Bayern, 80469, Deutschland
Altstadt-Lehel, Munich -> 35, Westermühlstraße, Bezirksteil Glockenbach, Ludwigsvorstadt-Isarvorstadt, München, Bayern, 80469, Deutschland
Altstadt-Lehel, Munich -> Erhardtstraße, Bezirksteil Deutsches Museum, Ludwigsvorstadt-Isarvorstadt, München, Bayern, 80469, Deutschland
Altstadt-Lehel, Munich -> 45, Zeppelinstraße, Untere Au, Bezirksteil Untere Au, Au, Au-Haidhausen, München, Bayern, 81669, Deutschland
Altstadt-Lehel, Munich -> 2, Haydnstraße, Klinikviertel, Bezirksteil Ludwigsvorstadt-Kliniken, Ludwigsvorstadt-Isarvorstadt, München, Bayern, 80336, Deutschland
Altstadt-Lehel, Munich -> 7, Maistraße, Bezirksteil Am alten südlichen Friedhof, Ludwigsvorstadt-Isarvorstadt, München, Bayern, 80337, Deutschland
A

So, in total we have 1204 venues from all grid bubbles. Lets check how many venues were returned from all of these addresses.

In [65]:
city_venues

Neighborhood  \
0    Altstadt-Lehel, Munich   
1    Altstadt-Lehel, Munich   
2    Altstadt-Lehel, Munich   
3    Altstadt-Lehel, Munich   
4    Altstadt-Lehel, Munich   
..                      ...   
776            Laim, Munich   
777            Laim, Munich   
778            Laim, Munich   
779            Laim, Munich   
780            Laim, Munich   

                                              District  District Latitude  \
0    24, Eduard-Schmid-Straße, Untere Au, Bezirkste...          48.124495   
1    35, Westermühlstraße, Bezirksteil Glockenbach,...          48.127743   
2    35, Westermühlstraße, Bezirksteil Glockenbach,...          48.127743   
3    35, Westermühlstraße, Bezirksteil Glockenbach,...          48.127743   
4    2, Haydnstraße, Klinikviertel, Bezirksteil Lud...          48.130646   
..                                                 ...                ...   
776  17, Agricolastraße, Bezirksteil St. Ulrich, La...          48.142620   
777  Edeka, 390, Landsberger Straße, Am Westbad, Be...          48.145690   
778  Edeka, 390, Landsberger Straße, Am Westbad, Be...          48.145690   
779  5, Schloßschmidstraße, Siedlung Neuhausen, Bez...          48.146741   
780  Königlicher Hirschgarten, Hirschgarten, Birket...          48.149812   

     District Longitude                        id  \
0             11.575470  4e3d0f8f45dd68e32733f731   
1             11.572374  4b1589b7f964a52063ae23e3   
2             11.572374  4ba3ee9df964a520dd6e38e3   
3             11.572374  51210c87e4b07e643351484f   
4             11.557778  523b5140498e98204df77776   
..                  ...                       ...   
776           11.493313  502155cee4b0f7aa0b19ae21   
777           11.484460  4e188808b0fb8567c66d6ed3   
778           11.484460  4beeb11ce24d20a13a327314   
779           11.518965  5565fe7d498e124cb64a7ef5   
780           11.510112  4dcbcbd6e4cde609a3a005e8   

                                  Venue  Venue Latitude  Venue Longitude  \
0                              Isarfest       48.124145        11.575537   
1                          Kirschbluete       48.127920        11.573120   
2                            Maroto Bar       48.127438        11.571582   
3                     Vintage Selection       48.127172        11.571875   
4           Hans im Glück - Burgergrill       48.130126        11.557854   
..                                  ...             ...              ...   
776                       Agricolaplatz       48.142236        11.492911   
777                               EDEKA       48.145099        11.484187   
778                            ALDI SÜD       48.145265        11.485066   
779                     Juli Restaurant       48.146195        11.518886   
780  Kulturpavillon München-Nymphenburg       48.149689        11.511064   

                                Venue Address        Venue Category  
0                                     München  Arts & Entertainment  
1                             Ickstattstr. 26      Asian Restaurant  
2                           Westermühlstr. 31                   Bar  
3                         Westermühlstraße 39              Wine Bar  
4                  Goetheplatz 2 (Mozartstr.)          Burger Joint  
..                                        ...                   ...  
776                             Agricolaplatz                 Plaza  
777                      Landsberger Str. 390           Supermarket  
778                      Landsberger Str. 388           Supermarket  
779                      Friesenheimer Brücke   Szechuan Restaurant  
780  Arnulfstraße 294 (Wendl-Dietrich-Straße)           Art Gallery  

[781 rows x 10 columns]

In [66]:
city_venues['Venue Category'].unique()

array(['Arts & Entertainment', 'Asian Restaurant', 'Bar', 'Wine Bar',
       'Burger Joint', 'Movie Theater', 'Cocktail Bar',
       'Doner Restaurant', 'Theater', 'Currywurst Joint',
       'Food & Drink Shop', 'Bistro', 'Hotel',
       'Vegetarian / Vegan Restaurant', 'Nightclub',
       'College Academic Building', 'Camera Store', 'Turkish Restaurant',
       'Bank', 'Museum', 'Jewish Restaurant', 'Plaza',
       'Falafel Restaurant', 'Fish Market', 'Seafood Restaurant', 'Café',
       'English Restaurant', 'Argentinian Restaurant',
       'German Restaurant', 'Bathing Area', 'Indie Movie Theater',
       'Cupcake Shop', 'Bavarian Restaurant', 'Clothing Store',
       'Japanese Restaurant', 'Park', 'Bakery', 'Trattoria/Osteria',
       'Coffee Shop', 'Gym', 'Bookstore', 'Pizza Place', 'Supermarket',
       'Palatine Restaurant', 'Deli / Bodega', 'Electronics Store',
       'Tunnel', 'Tea Room', 'Athletics & Sports', 'Lake',
       'Miscellaneous Shop', 'Pub', 'Dumpling Restaurant',


In [67]:
city_venues.to_csv('city_venues_munich.csv', index = False)

In [68]:
city_venues.groupby('Venue Category').count()

Neighborhood  District  District Latitude  \
Venue Category                                                      
Accessories Store                  1         1                  1   
Afghan Restaurant                  2         2                  2   
American Restaurant                2         2                  2   
Arcade                             1         1                  1   
Argentinian Restaurant             1         1                  1   
...                              ...       ...                ...   
Vietnamese Restaurant              6         6                  6   
Vineyard                           1         1                  1   
Wine Bar                           1         1                  1   
Wine Shop                          1         1                  1   
Zoo Exhibit                        7         7                  7   

                        District Longitude  id  Venue  Venue Latitude  \
Venue Category                                                          
Accessories Store                        1   1      1               1   
Afghan Restaurant                        2   2      2               2   
American Restaurant                      2   2      2               2   
Arcade                                   1   1      1               1   
Argentinian Restaurant                   1   1      1               1   
...                                    ...  ..    ...             ...   
Vietnamese Restaurant                    6   6      6               6   
Vineyard                                 1   1      1               1   
Wine Bar                                 1   1      1               1   
Wine Shop                                1   1      1               1   
Zoo Exhibit                              7   7      7               7   

                        Venue Longitude  Venue Address  
Venue Category                                          
Accessories Store                     1              1  
Afghan Restaurant                     2              2  
American Restaurant                   2              2  
Arcade                                1              1  
Argentinian Restaurant                1              1  
...                                 ...            ...  
Vietnamese Restaurant                 6              6  
Vineyard                              1              1  
Wine Bar                              1              1  
Wine Shop                             1              1  
Zoo Exhibit                           7              7  

[184 rows x 9 columns]

In [69]:
print('There are {} unique categories.'.format(len(city_venues['Venue Category'].unique())))

There are 184 unique categories.


There are 184 unique categories.

Analyze Each Property's surrounding area - using one hot coding, each venue category will become a column in the dataframe:

In [70]:
# one hot encoding
city_venues_onehot = pd.get_dummies(city_venues[['Venue Category']], prefix="", prefix_sep="")

# add property column back to dataframe
city_venues_onehot['Neighborhood'] = city_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [city_venues_onehot.columns[-1]] + list(city_venues_onehot.columns[:-1])
city_venues_onehot = city_venues_onehot[fixed_columns]

city_venues_onehot.head()

Neighborhood  Accessories Store  Afghan Restaurant  \
0  Altstadt-Lehel, Munich                  0                  0   
1  Altstadt-Lehel, Munich                  0                  0   
2  Altstadt-Lehel, Munich                  0                  0   
3  Altstadt-Lehel, Munich                  0                  0   
4  Altstadt-Lehel, Munich                  0                  0   

   American Restaurant  Arcade  Argentinian Restaurant  Art Gallery  \
0                    0       0                       0            0   
1                    0       0                       0            0   
2                    0       0                       0            0   
3                    0       0                       0            0   
4                    0       0                       0            0   

   Art Museum  Arts & Entertainment  Asian Restaurant  ...  Tram Station  \
0           0                     1                 0  ...             0   
1           0                     0                 1  ...             0   
2           0                     0                 0  ...             0   
3           0                     0                 0  ...             0   
4           0                     0                 0  ...             0   

   Trattoria/Osteria  Tunnel  Turkish Restaurant  \
0                  0       0                   0   
1                  0       0                   0   
2                  0       0                   0   
3                  0       0                   0   
4                  0       0                   0   

   Vegetarian / Vegan Restaurant  Vietnamese Restaurant  Vineyard  Wine Bar  \
0                              0                      0         0         0   
1                              0                      0         0         0   
2                              0                      0         0         0   
3                              0                      0         0         1   
4                              0                      0         0         0   

   Wine Shop  Zoo Exhibit  
0          0            0  
1          0            0  
2          0            0  
3          0            0  
4          0            0  

[5 rows x 185 columns]

In [71]:
city_venues_onehot.shape

(781, 185)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category:

In [72]:
city_venues_grouped = city_venues_onehot.groupby('Neighborhood').mean().reset_index()
city_venues_grouped

Neighborhood  Accessories Store  \
0                Allach Untermenzing, Munich           0.000000   
1                     Altstadt-Lehel, Munich           0.000000   
2                      Au-Haidhausen, Munich           0.000000   
3         Aubing-Lochhausen-Langwied, Munich           0.000000   
4                       Berg am Laim, Munich           0.000000   
5                        Bogenhausen, Munich           0.000000   
6             Feldmoching-Hasenbergl, Munich           0.000000   
7                        Forstenried, Munich           0.000000   
8                  Fürstenried-Solln, Munich           0.000000   
9                             Hadern, Munich           0.000000   
10                              Laim, Munich           0.000000   
11  Ludwigsvorstadt and Isarvorstadt, Munich           0.000000   
12                       Maxvorstadt, Munich           0.000000   
13         Milbertshofen und Am Hart, Munich           0.000000   
14                           Moosach, Munich           0.000000   
15             Neuhausen Nymphenburg, Munich           0.000000   
16                       Obergiesing, Munich           0.000000   
17                      Obersendling, Munich           0.000000   
18              Pasing – Obermenzing, Munich           0.000000   
19            Ramersdorf und Perlach, Munich           0.000000   
20                    Schwabing West, Munich           0.000000   
21                Schwabing-Freimann, Munich           0.000000   
22                  Schwanthalerhöhe, Munich           0.000000   
23               Sendling – Westpark, Munich           0.000000   
24                          Sendling, Munich           0.000000   
25                       Thalkirchen, Munich           0.000000   
26                  Trudering – Riem, Munich           0.000000   
27       Untergiesing und Harlaching, Munich           0.026316   

    Afghan Restaurant  American Restaurant    Arcade  Argentinian Restaurant  \
0            0.000000             0.000000  0.000000                0.000000   
1            0.000000             0.000000  0.000000                0.010989   
2            0.000000             0.000000  0.000000                0.000000   
3            0.000000             0.000000  0.000000                0.000000   
4            0.000000             0.000000  0.000000                0.000000   
5            0.000000             0.000000  0.000000                0.000000   
6            0.000000             0.000000  0.000000                0.000000   
7            0.000000             0.000000  0.000000                0.000000   
8            0.000000             0.000000  0.000000                0.000000   
9            0.000000             0.000000  0.000000                0.000000   
10           0.000000             0.000000  0.000000                0.000000   
11           0.011765             0.000000  0.000000                0.000000   
12           0.000000             0.000000  0.000000                0.000000   
13           0.000000             0.000000  0.000000                0.000000   
14           0.000000             0.133333  0.000000                0.000000   
15           0.055556             0.000000  0.000000                0.000000   
16           0.000000             0.000000  0.000000                0.000000   
17           0.000000             0.000000  0.000000                0.000000   
18           0.000000             0.000000  0.000000                0.000000   
19           0.000000             0.000000  0.000000                0.000000   
20           0.000000             0.000000  0.000000                0.000000   
21           0.000000             0.000000  0.000000                0.000000   
22           0.000000             0.000000  0.000000                0.000000   
23           0.000000             0.000000  0.000000                0.000000   
24           0.000000             0.000000  0.000000                0.000000   
25      

In [73]:
city_venues_grouped.shape

(28, 185)

In [74]:
num_top_venues = 7

for hood in city_venues_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = city_venues_grouped[city_venues_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Allach Untermenzing, Munich----
                        venue  freq
0                  Playground  0.25
1         Sporting Goods Shop  0.25
2                      Forest  0.25
3                 Snack Place  0.25
4          Mexican Restaurant  0.00
5          Miscellaneous Shop  0.00
6  Modern European Restaurant  0.00


----Altstadt-Lehel, Munich----
                 venue  freq
0                 Café  0.07
1                Hotel  0.05
2                  Bar  0.05
3     Asian Restaurant  0.04
4          Supermarket  0.03
5         Burger Joint  0.03
6  Indie Movie Theater  0.03


----Au-Haidhausen, Munich----
                   venue  freq
0                   Café  0.09
1     Italian Restaurant  0.06
2                  Hotel  0.04
3      Indian Restaurant  0.04
4           Burger Joint  0.04
5  Vietnamese Restaurant  0.04
6       Doner Restaurant  0.04


----Aubing-Lochhausen-Langwied, Munich----
                        venue  freq
0           Recreation Center  0.33
1             

In [75]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [76]:
num_top_venues = 7

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
city_venues_sorted = pd.DataFrame(columns=columns)
city_venues_sorted['Neighborhood'] = city_venues_grouped['Neighborhood']

for ind in np.arange(city_venues_grouped.shape[0]):
    city_venues_sorted.iloc[ind, 1:] = return_most_common_venues(city_venues_grouped.iloc[ind, :], num_top_venues)

city_venues_sorted

Neighborhood 1st Most Common Venue  \
0                Allach Untermenzing, Munich            Playground   
1                     Altstadt-Lehel, Munich                  Café   
2                      Au-Haidhausen, Munich                  Café   
3         Aubing-Lochhausen-Langwied, Munich        Clothing Store   
4                       Berg am Laim, Munich              Bus Stop   
5                        Bogenhausen, Munich              Bus Stop   
6             Feldmoching-Hasenbergl, Munich          Cocktail Bar   
7                        Forstenried, Munich              Bus Stop   
8                  Fürstenried-Solln, Munich                Bakery   
9                             Hadern, Munich              Bus Stop   
10                              Laim, Munich           Supermarket   
11  Ludwigsvorstadt and Isarvorstadt, Munich                 Hotel   
12                       Maxvorstadt, Munich                   Bar   
13         Milbertshofen und Am Hart, Munich           Supermarket   
14                           Moosach, Munich                Bakery   
15             Neuhausen Nymphenburg, Munich      Greek Restaurant   
16                       Obergiesing, Munich                Bakery   
17                      Obersendling, Munich                 Hotel   
18              Pasing – Obermenzing, Munich                Bakery   
19            Ramersdorf und Perlach, Munich              Bus Stop   
20                    Schwabing West, Munich           Supermarket   
21                Schwabing-Freimann, Munich            Restaurant   
22                  Schwanthalerhöhe, Munich    Italian Restaurant   
23               Sendling – Westpark, Munich                  Café   
24                          Sendling, Munich    Italian Restaurant   
25                       Thalkirchen, Munich           Zoo Exhibit   
26                  Trudering – Riem, Munich              Bus Stop   
27       Untergiesing und Harlaching, Munich              Bus Stop   

   2nd Most Common Venue 3rd Most Common Venue       4th Most Common Venue  \
0    Sporting Goods Shop           Snack Place                      Forest   
1                  Hotel                   Bar            Asian Restaurant   
2     Italian Restaurant     Indian Restaurant            Doner Restaurant   
3      Recreation Center              Bus Stop                Dance Studio   
4            Supermarket                 Hotel  Construction & Landscaping   
5              Drugstore                Bakery          Chinese Restaurant   
6            IT Services       Motorcycle Shop                  Playground   
7                 Bakery             BBQ Joint                      Castle   
8         Farmers Market           Supermarket                 Pizza Place   
9      Trattoria/Osteria           Supermarket          Salon / Barbershop   
10      Doner Restaurant                 Plaza          Light Rail Station   
11                  Café                   Bar          Italian Restaurant   
12                  Café    Italian Restaurant            Asian Restaurant   
13              Bus Stop          Tram Station                  Restaurant   
14   American Restaurant    Chinese Restaurant            Greek Restaurant   
15     Trattoria/Osteria    Italian Restaurant                Cocktail Bar   
16                  Park           Supermarket        Gym / Fitness Center   
17              Bus Stop           Zoo Exhibit          Athletics & Sports   
18            Steakhouse              Bus Stop          Italian Restaurant   
19           Supermarket         Garden Center                   Drugstore   
20                Bakery        Breakfast Spot                Tram Station   
21                Bistro                  Park        Fast Food Restaurant   
22                 Hotel            Playground                 Pizza Place   
23      Greek Restaurant           Supermarket                 Flower Shop   
24            Playground                  Café     

In [77]:
# set number of clusters
kclusters = 5

city_venues_grouped_clustering = city_venues_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(city_venues_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([4, 1, 1, 2, 0, 0, 3, 0, 0, 0])

In [78]:
# add clustering labels
city_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

city_venues_merged = df_locations

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
city_venues_merged = city_venues_merged.join(city_venues_sorted.set_index('Neighborhood'), on='Neighborhood')


city_venues_merged # check the last columns!

Neighborhood  \
0     Altstadt-Lehel, Munich   
1     Altstadt-Lehel, Munich   
2     Altstadt-Lehel, Munich   
3     Altstadt-Lehel, Munich   
4     Altstadt-Lehel, Munich   
...                      ...   
1199            Laim, Munich   
1200            Laim, Munich   
1201            Laim, Munich   
1202            Laim, Munich   
1203            Laim, Munich   

                                                Address   Latitude  Longitude  \
0     24, Eduard-Schmid-Straße, Untere Au, Bezirkste...  48.124495  11.575470   
1     58, Pestalozzistraße, Bezirksteil Glockenbach,...  48.127571  11.566624   
2     35, Westermühlstraße, Bezirksteil Glockenbach,...  48.127743  11.572374   
3     Erhardtstraße, Bezirksteil Deutsches Museum, L...  48.127914  11.578123   
4     45, Zeppelinstraße, Untere Au, Bezirksteil Unt...  48.128086  11.583872   
...                                                 ...        ...        ...   
1199  An der Schloßmauer, Bezirksteil Nymphenburg, N...  48.149286  11.492859   
1200  Zuccalistraße, Bezirksteil Nymphenburg, Neuhau...  48.149462  11.498610   
1201  30, Wotanstraße, Bezirksteil Nymphenburg, Neuh...  48.149637  11.504361   
1202  Königlicher Hirschgarten, Hirschgarten, Birket...  48.149812  11.510112   
1203  57, Richildenstraße, Bezirksteil Nymphenburg, ...  48.152883  11.501259   

                  X             Y  Distance from center  Cluster Labels  \
0     245177.072223  5.335811e+06           1484.614978               1   
1     244534.215080  5.336182e+06           1285.714286               1   
2     244962.786509  5.336182e+06           1133.893419               1   
3     245391.357937  5.336182e+06           1133.893419               1   
4     245819.929366  5.336182e+06           1285.714286               1   
...             ...           ...                   ...             ...   
1199  239156.457330  5.338843e+06           1285.714286               0   
1200  239585.028758  5.338843e+06           1133.893419               0   
1201  240013.600187  5.338843e+06           1133.893419               0   
1202  240442.171616  5.338843e+06           1285.714286               0   
1203  239799.314473  5.339214e+06           1484.614978               0   

     1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                     Café                 Hotel                   Bar   
1                     Café                 Hotel                   Bar   
2                     Café                 Hotel                   Bar   
3                     Café                 Hotel                   Bar   
4                     Café                 Hotel                   Bar   
...                    ...                   ...                   ...   
1199           Supermarket      Doner Restaurant                 Plaza   
1200           Supermarket      Doner Restaurant                 Plaza   
1201           Supermarket      Doner Restaurant                 Plaza   
1202           Supermarket      Doner Restaurant                 Plaza   
1203           Supermarket      Doner Restaurant                 Plaza   

     4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0         Asian Restaurant   Indie Movie Theater           Supermarket   
1         Asian Restaurant   Indie Movie Theater           Supermarket   
2         Asian Restaurant   Indie Movie Theater           Supermarket   
3         Asian Restaurant   Indie Movie Theater           Supermarket   
4         Asian Restaurant   Indie Movie Theater           Supermarket   
...                    ...                   ...                   ...   
1199    Light Rail Station          Tram Station   Szechuan Restaurant   
1200    Light Rail Station          Tram Station   Szechuan Restaurant   
1201    Light Rail Station          Tram Station   Szechuan Restaurant   
1202    Light Rail Station          Tram Station   Szechuan Restaurant   
1203    Light Rail Station          Tram Station   Sze

In [79]:
# create map
map_clusters = folium.Map(location=dfListing['location_corr'][1], zoom_start=10.4)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(city_venues_merged['Latitude'], city_venues_merged['Longitude'], city_venues_merged['Neighborhood'], city_venues_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.Circle(
        [lat, lon],
        radius=m/bubble_detail_factor,
        popup=label,
        color=rainbow[cluster-1],
        fill=False,
        fill_color=rainbow[cluster-1]).add_to(map_clusters)
       
map_clusters

In [80]:
examine_clusters = city_venues_merged.drop(['Address','Latitude','Longitude','X','Y','Distance from center'],1).drop_duplicates()
examine_clusters

Neighborhood  Cluster Labels  \
0                       Altstadt-Lehel, Munich               1   
43    Ludwigsvorstadt and Isarvorstadt, Munich               1   
86                         Maxvorstadt, Munich               1   
129                     Schwabing West, Munich               1   
172                      Au-Haidhausen, Munich               1   
215                           Sendling, Munich               1   
258                Sendling – Westpark, Munich               1   
301                   Schwanthalerhöhe, Munich               1   
344              Neuhausen Nymphenburg, Munich               1   
387                            Moosach, Munich               1   
430          Milbertshofen und Am Hart, Munich               0   
473                 Schwabing-Freimann, Munich               1   
516                        Bogenhausen, Munich               0   
559                       Berg am Laim, Munich               0   
602                   Trudering – Riem, Munich               0   
645             Ramersdorf und Perlach, Munich               0   
688                        Obergiesing, Munich               1   
731        Untergiesing und Harlaching, Munich               1   
774                        Thalkirchen, Munich               1   
817                       Obersendling, Munich               1   
860                        Forstenried, Munich               0   
903                  Fürstenried-Solln, Munich               0   
946                             Hadern, Munich               0   
989               Pasing – Obermenzing, Munich               1   
1032        Aubing-Lochhausen-Langwied, Munich               2   
1075               Allach Untermenzing, Munich               4   
1118            Feldmoching-Hasenbergl, Munich               3   
1161                              Laim, Munich               0   

     1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                     Café                 Hotel                   Bar   
43                   Hotel                  Café                   Bar   
86                     Bar                  Café    Italian Restaurant   
129            Supermarket                Bakery        Breakfast Spot   
172                   Café    Italian Restaurant     Indian Restaurant   
215     Italian Restaurant            Playground                  Café   
258                   Café      Greek Restaurant           Supermarket   
301     Italian Restaurant                 Hotel            Playground   
344       Greek Restaurant     Trattoria/Osteria    Italian Restaurant   
387                 Bakery   American Restaurant    Chinese Restaurant   
430            Supermarket              Bus Stop          Tram Station   
473             Restaurant                Bistro                  Park   
516               Bus Stop             Drugstore                Bakery   
559               Bus Stop           Supermarket                 Hotel   
602               Bus Stop                Hostel       Organic Grocery   
645               Bus Stop           Supermarket         Garden Center   
688                 Bakery                  Park           Supermarket   
731               Bus Stop                 Beach                  Café   
774            Zoo Exhibit              Bus Stop                 Hotel   
817                  Hotel              Bus Stop           Zoo Exhibit   
860               Bus Stop                Bakery             BBQ Joint   
903                 Bakery        Farmers Market           Supermarket   
946               Bus Stop     Trattoria/Osteria           Supermarket   
989                 Bakery            Steakhouse              Bus Stop   
1032        Clothing Store     Recreation Center              Bus Stop   
1075            Playground   Sporting Goods Shop           Snack Place   
1118          Cocktail Bar           IT Services       Motorcycle Shop   
1161           Supermarket      Doner Restauran

In [81]:
examine_clusters.loc[examine_clusters['Cluster Labels'] == 0, examine_clusters.columns[[0] + list(range(2, examine_clusters.shape[1]))]]

Neighborhood 1st Most Common Venue  \
430   Milbertshofen und Am Hart, Munich           Supermarket   
516                 Bogenhausen, Munich              Bus Stop   
559                Berg am Laim, Munich              Bus Stop   
602            Trudering – Riem, Munich              Bus Stop   
645      Ramersdorf und Perlach, Munich              Bus Stop   
860                 Forstenried, Munich              Bus Stop   
903           Fürstenried-Solln, Munich                Bakery   
946                      Hadern, Munich              Bus Stop   
1161                       Laim, Munich           Supermarket   

     2nd Most Common Venue 3rd Most Common Venue       4th Most Common Venue  \
430               Bus Stop          Tram Station                  Restaurant   
516              Drugstore                Bakery          Chinese Restaurant   
559            Supermarket                 Hotel  Construction & Landscaping   
602                 Hostel       Organic Grocery         Rental Car Location   
645            Supermarket         Garden Center                   Drugstore   
860                 Bakery             BBQ Joint                      Castle   
903         Farmers Market           Supermarket                 Pizza Place   
946      Trattoria/Osteria           Supermarket          Salon / Barbershop   
1161      Doner Restaurant                 Plaza          Light Rail Station   

     5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  
430                   Bank                Bakery                Bus Station  
516                Dog Run           Supermarket  Middle Eastern Restaurant  
559           Tram Station            Restaurant                 Theme Park  
602                   Café     Food & Drink Shop         Italian Restaurant  
645                Dog Run         Shopping Mall             Clothing Store  
860                Dog Run           Supermarket           Asian Restaurant  
903          Bowling Alley                 Plaza         Mexican Restaurant  
946               Bus Line                Forest               Soccer Field  
1161          Tram Station   Szechuan Restaurant                   Bus Stop

#### Reference:

https://www.linkedin.com/pulse/finding-surrounding-venues-before-buying-home-tony-chow/